api/dimensions.go

```diff
@@ -187,22 +217,28 @@ func (api *DatasetAPI) getDimensionOptions(w http.ResponseWriter, r *http.Reques
 	log.Event(ctx, "get dimension options", log.INFO, logData)
 }
 
-func handleDimensionsErr(ctx context.Context, w http.ResponseWriter, err error, data log.Data) {
+// handleDimensionsErr maps the provided error to its corresponding status code.
+// The error is logged with ERROR severity, but the stack trace is only shown for errors corresponding to InternalServerError status
+func handleDimensionsErr(ctx context.Context, w http.ResponseWriter, msg string, err error, data log.Data) {
 	if data == nil {
 		data = log.Data{}
 	}
 
-	var status int
-	response := err
 	switch {
+	case errs.BadRequestMap[err]:
+		data["response_status"] = http.StatusBadRequest
+		data["user_error"] = err.Error()
+		log.Event(ctx, fmt.Sprintf("request unsuccessful: %s", msg), log.ERROR, data)
+		http.Error(w, err.Error(), http.StatusBadRequest)
```

Comment:
Each case is duplicating the same code steps. Consider assigning variables in each of case clauses and then doing error response and logging once at the end using those vars. It would remove a lot of the duplication and make it easier to extend in future.

Something like:

```
var status int
var msg string
switch {
    case errs.NotFoundMap[err]:
        status =  http.StatusNotFound
        msg = err.Error()
    default:
        ...
}
data["response_status"] = status
log.Event(ctx, fmt.Sprintf("request unsuccessful: %s", msg), log.ERROR, log.Error(err), data)
http.Error(w, msg, status)
```

Reply:
That wont work because log.Error is only done in the default case so as to only do stack dump traces when needed

Reply:
I don't think there is any harm in logging the error in all cases?

Reply:
The code does return the 'user' error in:
data["user_error"] = err.Error()
and in specific cases where it is only a user error then the code does no longer do a stack trace.
All other errors still get a stack trace.
For example if the user enters a negative 'limit' for this block of code, its a user error not a program or database type error.

Reply:
@redhug1  suggestion was to still log the error, but add the error string as a logData paramter only, without a stack trace, for 'user-errors' like validation, not found, etc:
```
data["user_error"] = err.Error()
```
For any error that results in 500 response, we would still log the stack trace, as it might be something unexpected

Reply:
Example of response using this:
```
{
  "created_at": "2020-12-01T15:04:04.643317Z",
  "data": {
    "authenticated": true,
    "caller_identity": "florence@magicroundabout.ons.gov.uk",
    "dataset_id": "cpih01",
    "dimension": "aggregate",
    "edition": "time-series",
    "func": "getDimensionOptions",
    "query_params": "offset=88w",
    "response_status": 400,
    "user_error": "invalid query parameter",
    "user_identity": "florence@magicroundabout.ons.gov.uk",
    "version": "1"
  },
  "event": "request unsuccessful: failed to obtain offset from request query paramters",
  "namespace": "dp-dataset-api",
  "severity": 1,
  "trace_id": "gtwQOYuarlyuIQbw"
}
```

Same log error with stack trace:

```
{
  "created_at": "2020-12-01T15:04:59.766085Z",
  "data": {
    "authenticated": true,
    "caller_identity": "florence@magicroundabout.ons.gov.uk",
    "dataset_id": "cpih01",
    "dimension": "aggregate",
    "edition": "time-series",
    "func": "getDimensionOptions",
    "query_params": "offset=88w",
    "response_status": 400,
    "user_error": "invalid query parameter",
    "user_identity": "florence@magicroundabout.ons.gov.uk",
    "version": "1"
  },
  "error": {
    "data": {},
    "error": "invalid query parameter",
    "stack_trace": [
      {
        "file": "/Users/david/gomod/src/github.com/ONSdigital/dp-dataset-api/api/dimensions.go",
        "function": "github.com/ONSdigital/dp-dataset-api/api.handleDimensionsErr",
        "line": 231
      },
      {
        "file": "/Users/david/gomod/src/github.com/ONSdigital/dp-dataset-api/api/dimensions.go",
        "function": "github.com/ONSdigital/dp-dataset-api/api.(*DatasetAPI).getDimensionOptions",
        "line": 172
      },
      {
        "file": "/usr/local/Cellar/go@1.13/1.13.12/libexec/src/net/http/server.go",
        "function": "net/http.HandlerFunc.ServeHTTP",
        "line": 2007
      },
      {
        "file": "/Users/david/go/pkg/mod/github.com/!o!n!sdigital/dp-authorisation@v0.1.0/auth/nop.go",
        "function": "github.com/ONSdigital/dp-authorisation/auth.(*NopHandler).Require.func1",
        "line": 19
      },
      {
        "file": "/usr/local/Cellar/go@1.13/1.13.12/libexec/src/net/http/server.go",
        "function": "net/http.HandlerFunc.ServeHTTP",
        "line": 2007
      },
      {
        "file": "/Users/david/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go",
        "function": "github.com/gorilla/mux.(*Router).ServeHTTP",
        "line": 210
      },
      {
        "file": "/Users/david/go/pkg/mod/github.com/!o!n!sdigital/dp-net@v1.0.10/handlers/handlers.go",
        "function": "github.com/ONSdigital/dp-net/handlers.DoCheckHeader.func1",
        "line": 27
      },
      {
        "file": "/usr/local/Cellar/go@1.13/1.13.12/libexec/src/net/http/server.go",
        "function": "net/http.HandlerFunc.ServeHTTP",
        "line": 2007
      },
      {
        "file": "/Users/david/go/pkg/mod/github.com/!o!n!sdigital/dp-net@v1.0.10/handlers/identity.go",
        "function": "github.com/ONSdigital/dp-net/handlers.identityWithHTTPClient.func1.1",
        "line": 64
      },
      {
        "file": "/usr/local/Cellar/go@1.13/1.13.12/libexec/src/net/http/server.go",
        "function": "net/http.HandlerFunc.ServeHTTP",
        "line": 2007
      }
    ]
  },
  "event": "request unsuccessful: failed to obtain offset from request query paramters",
  "namespace": "dp-dataset-api",
  "severity": 1,
  "trace_id": "UVVSUIIutGhIQstB"
}
```

Reply:
Alternatively, I can go back to @redhug1  original suggestion and take the logging out of this function, and only add the `log.Error(err)` where needed. But the problem with this approach is that we would end up doing error checking in lots of places for logging purposes, and then calling the function that is supposed to do the error handling (in order to set the corresponding status code)